File: src/10.graph/40.433.ts
export function minMutation(
    startGene: string,
    endGene: string,
    bank: string[]
): number {
    const bankSet = new Set(bank);
    const geneList = ["A", "C", "G", "T"];

    if (!bankSet.has(endGene)) {
        return -1;
    }

    const queue: string[] = [startGene];
    let count = 1;
    while (queue.length) {
        const size = queue.length;
        for (let i = 0; i < size; i++) {
            const curGene = queue.shift();

            for (let j = 0; j < curGene.length; j++) {
                for (let k = 0; k < geneList.length; k++) {
                    if (geneList[k] === curGene[j]) {
                        continue;
                    }
                    const newGene =
                        curGene.slice(0, j) +
                        geneList[k] +
                        curGene.slice(j + 1);

                    if (newGene === endGene) {
                        return count;
                    }

                    if (bankSet.has(newGene)) {
                        bankSet.delete(newGene);
                        queue.push(newGene);
                    }
                }
            }
        }

        count++;
    }

    return -1;
}

File: src/10.graph/40.433.go
package graph

func MinMutation(startGene string, endGene string, bank []string) int {
	geneList := []byte{'A', 'C', 'G', 'T'}
	bankSet := make(map[string]bool)

	for _, v := range bank {
		bankSet[v] = true
	}

	if !bankSet[endGene] {
		return -1
	}

	queue := []string{startGene}
	count := 1

	for len(queue) > 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			curGene := queue[0]
			queue = queue[1:]

			for j := 0; j < 8; j++ {
				for k := 0; k < 4; k++ {
					if geneList[k] == curGene[j] {
						continue
					}

					newGene := curGene[0:j] + string(geneList[k]) + curGene[j+1:]

					if newGene == endGene {
						return count
					}

					if bankSet[newGene] {
						bankSet[newGene] = false
						queue = append(queue, newGene)
					}
				}
			}

		}
		count++
	}

	return -1
}
